modules/programs/thunderbird.nix

```diff
@@ -0,0 +1,267 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.programs.thunderbird;
+
+  enabledAccounts = attrValues
+    (filterAttrs (_: a: a.thunderbird.enable) config.accounts.email.accounts);
+
+  enabledAccountsWithId =
+    imap1 (i: v: v // { id = toString i; }) enabledAccounts;
```

Comment:
Adding and removing accounts can change the ids, but it wouldn't *directly* be a problem for enabled accounts, because `mail.accountmanager.accounts` would be set to account for the change.

The problem is that there is no way to make sure `prefs.js` does not contain any leftover config from a previous generation. So adding a new account might "reactivate" preferences that were meant to be used with a previous account with the same id, for example. But this can also happen with explicit ids (removing account 2, then adding a new account with id 2). The Firefox config has similar issues.

I'm fine with switching to explicit account identifiers, if you think it would be better. Maybe we should just have a warning explaining that changes to accounts might require deleting `prefs.js`.

Reply:
Now that I think about it, it might be possible to avoid "reactivating" preferences by using some hash as account identifier (using `builtins.hashString` with selected fields). Might be somewhat overengineered, though.

Reply:
I would prefer to get rid of the id (as just ordinal number).

Hash seems better to me than just ordinal in current order.

Is explicit tag needed for `accounts.email.accounts...settings` parameter or can it just use the account name?

I just use accounts with explicit names, no numbers like `mail.server.server_aalto` `mail.identity.id_aalto`.

ps. To me it seems thunderbird and firefox are not really optimal to be used like this in general terms, it seems to me that prefs.js overrides everything, even lockPref settings, so it's very hard to give almost any guarantees to the end user. I did notice eg. that thunderbird does change directory settings, if a directory is temporarily unavailable due to other issues. I would still like to try, in case we find a better way in the future.

Reply:
I was initially using account names as identifiers, but the problem was that accounts can have essentially any name, possibly including dots or any character that could break preferences (e.g., email addresses, for which I think any UTF-8 character is allowed?). A potential solution would have been percent-encoding the name, since Thunderbird already generates and parses percent-encoded names by itself in some cases, but there is no builtin or library function to do this, and I wasn't able to find a simple, clean way to implement it.

Reply:
I guess that's not very much value over sha256, while it would look nicer.